### PR TITLE
[inductor] weight prepack for _convolution_transpose_pointwise

### PR DESCRIPTION
**Pull request by chunyuan-w**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90267
* __->__ #90266
* #90265
* #90264

This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 1]
```diff
@@ -643,22 +647,42 @@ Tensor _mkldnn_convolution_transpose(
...
+  } else {
+    // aten tensor with the shape of the prepacked tensor
+    at::Tensor origin_weight_t;
+    if (groups > 1) {
+      origin_weight_t = weight.transpose(0, 1).reshape(weight_IOHW_sizes);
+    } else {
+      origin_weight_t = weight.transpose(0, 1);
+    }
+    w = itensor_from_tensor(origin_weight_t);
+    w.transpose_(0, 1);
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:677]: Do we have a case that we want to handle this "aten tensor with the shape of the prepacked tensor", or we always model prepacked weight with "MKLDNNTensor"?

Comment (chunyuan-w) [reply aten/src/ATen/native/mkldnn/Conv.cpp:677]: When compiling the graph, each node is run with FakeTensor inputs and in this case, the weight tensor is an aten tensor with the shape of the prepacked tensor.

Comment (chunyuan-w) [reply aten/src/ATen/native/mkldnn/Conv.cpp]: For Meta tensor: added an implementation `mkldnn_convolution_transpose_pointwise_meta`.

For MKLDNN and CPU tensor: supported the computation for both of them in the original implementation: `mkldnn_convolution_transpose_pointwise`.

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 2]
```diff
@@ -643,22 +648,33 @@ Tensor _mkldnn_convolution_transpose(
...
-  ideep::tensor w = itensor_from_tensor(weight_t);
-  // mkldnn transposed convolution has weight in logical order of OIHW or OIDHW,
-  // while PyTorch has IOHW or IODHW, `._tranpose()` switches strides (no memory copy).
-  w.transpose_(0, 1);
+  auto output_sizes = conv_input_size(input.sizes(), weight_IOHW_sizes, padding, output_padding, stride, dilation, groups);
+  auto output = at::empty({0}, input.options());
+
+  const ideep::tensor x = itensor_from_tensor(input);
+
+  ideep::tensor w = itensor_from_tensor(weight);;
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:665]:
```suggestion
  ideep::tensor w = itensor_from_tensor(weight);
```

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 3]
```diff
@@ -709,6 +725,8 @@ Tensor mkldnn_convolution_transpose_pointwise(
     torch::List<c10::optional<at::Scalar>> scalars,
     c10::optional<c10::string_view> algorithm) {
   c10::impl::ExcludeDispatchKeyGuard edkg(c10::autograd_dispatch_keyset);
+  bool use_channels_last =
+      weight_t.is_mkldnn() || mkldnn_conv_use_channels_last(input_t, weight_t);
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:729]: The logic is ok. Not related to this PR, but does it make more sense that `mkldnn_conv_use_channels_last` should prefer channels last if input is a strided tensor and weight is mkldnn? @XiaobingSuper

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -3349,6 +3349,53 @@ def get_template_tiling(self):
         )
 
 
+# Port from aten/src/ATen/native/ConvUtils.h: _conv_input_size
+def _conv_input_size(
+    output_size, weight_size, padding, output_padding, stride, dilation, groups
+):
```
Comment (jgong5) [inline torch/_inductor/ir.py:3355]: How about we keep this function local to `_prepare_convolution_fusion_create` first? It is only used by this function.

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -3349,6 +3349,53 @@ def get_template_tiling(self):
...
+
+
+# The size of prepacked_weight is the prepacked size.
+#   Groups > 1:  [g*o, i/g, ...]
+#   Groups == 1: [o, i, ...]
+# Returns original weight size in [i, o, ...]
+def _original_deconv_weight_size(
+    prepacked_weight,
+    groups,
+):
```
Comment (jgong5) [inline torch/_inductor/ir.py:3384]: ditto.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mkldnn/Conv.cpp (hunk 2 above)
Comment (chunyuan-w) [reply aten/src/ATen/native/mkldnn/Conv.cpp]: Fixed the typo.

File: torch/_inductor/ir.py (hunk 4 above)
Comment (chunyuan-w) [reply torch/_inductor/ir.py]: Made this function local to `_prepare_convolution_fusion_create`.

File: torch/_inductor/ir.py (hunk 5 above)
Comment (chunyuan-w) [reply torch/_inductor/ir.py]: Fixed.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Comment by chunyuan-w** [comment]
Re-opened in https://github.com/pytorch/pytorch/pull/91955

**Pull request closed without merging**